plugin/evm/vmsync/client.go

```diff
@@ -169,52 +163,82 @@ func (client *client) stateSync(ctx context.Context) error {
 	// Register block syncer.
-	syncer, err := client.createBlockSyncer(client.summary.GetBlockHash(), client.summary.Height())
+	blockSyncer, err := client.createBlockSyncer(client.summary.GetBlockHash(), client.summary.Height())
 	if err != nil {
 		return fmt.Errorf("failed to create block syncer: %w", err)
 	}
-	if err := registry.Register(blockSyncOperationName, syncer); err != nil {
-		return fmt.Errorf("failed to register block syncer: %w", err)
+
+	codeQueue, err := client.createCodeQueue()
+	if err != nil {
+		return fmt.Errorf("failed to create code queue: %w", err)
 	}
 
-	// Register EVM state syncer.
-	evmSyncer, err := client.createEVMSyncer()
+	codeSyncer, err := client.createCodeSyncer(codeQueue.CodeHashes())
 	if err != nil {
-		return fmt.Errorf("failed to create EVM syncer: %w", err)
+		return fmt.Errorf("failed to create code syncer: %w", err)
 	}
 
-	if err := registry.Register(evmStateSyncOperationName, evmSyncer); err != nil {
-		return fmt.Errorf("failed to register EVM syncer: %w", err)
+	stateSyncer, err := client.createEVMSyncer(codeQueue)
+	if err != nil {
+		return fmt.Errorf("failed to create EVM state syncer: %w", err)
 	}
 
-	// Register atomic syncer.
+	var atomicSyncer syncpkg.Syncer
 	if client.Extender != nil {
-		atomicSyncer, err := client.createAtomicSyncer()
+		atomicSyncer, err = client.createAtomicSyncer()
 		if err != nil {
 			return fmt.Errorf("failed to create atomic syncer: %w", err)
 		}
+	}
 
-		if err := registry.Register(atomicStateSyncOperationName, atomicSyncer); err != nil {
-			return fmt.Errorf("failed to register atomic syncer: %w", err)
+	syncers := []syncpkg.Syncer{
+		blockSyncer,
+		codeSyncer,
+		stateSyncer,
+	}
+	if atomicSyncer != nil {
+		syncers = append(syncers, atomicSyncer)
+	}
+
+	for _, s := range syncers {
+		if err := registry.Register(s); err != nil {
+			return fmt.Errorf("failed to register %s syncer: %w", s.Name(), err)
 		}
 	}
 
 	return nil
 }
 
-func (client *client) createBlockSyncer(fromHash common.Hash, fromHeight uint64) (synccommon.Syncer, error) {
+func (client *client) createBlockSyncer(fromHash common.Hash, fromHeight uint64) (syncpkg.Syncer, error) {
 	return blocksync.NewSyncer(client.Client, client.ChainDB, blocksync.Config{
 		FromHash:      fromHash,
 		FromHeight:    fromHeight,
 		BlocksToFetch: BlocksToFetch,
 	})
 }
 
-func (client *client) createEVMSyncer() (synccommon.Syncer, error) {
-	return statesync.NewSyncer(client.Client, client.ChainDB, client.summary.GetBlockRoot(), statesync.NewDefaultConfig(client.RequestSize))
+func (client *client) createEVMSyncer(fetcher syncpkg.CodeRequestQueue) (syncpkg.Syncer, error) {
+	return statesync.NewSyncer(
+		client.Client,
+		client.ChainDB,
+		client.summary.GetBlockRoot(),
+		fetcher,
+		statesync.NewDefaultConfig(client.RequestSize),
+	)
+}
+
+func (client *client) createCodeQueue() (syncpkg.CodeRequestQueue, error) {
+	return statesync.NewCodeQueue(
+		client.ChainDB,
+		client.StateSyncDone,
```

Comment:
(rant) This is badly named since it's actually a quit channel. "Done" implies that everything has finished cleanly, not that things should quit early. It's the root of many of the problems we've had to work around here.